python/pyarrow/array.pxi

```diff
@@ -1883,6 +1917,28 @@ cdef class Array(_PandasConvertible):
         device = GetResultValue(ExportDevice(self.sp_array))
         return device.device_type, device.device_id
 
+    @property
+    def device_type(self):
+        """
+        The device type where the array resides.
+
+        Returns
+        -------
+        DeviceAllocationType
+        """
+        return _wrap_device_allocation_type(self.sp_array.get().device_type())
+
+    @property
+    def is_cpu(self):
+        """
+        Whether the array is CPU-accessible.
+        """
+        return self.device_type == DeviceAllocationType.CPU
+
+    def _assert_cpu(self):
+        if not self.is_cpu:
```

Comment:
If we would want to speed this up a bit (it's called in many places, although typically should only give tiny overhead), you could also inline similar code as you used above for the python attributes, like `if self.sp_array.get().device_type() != CDeviceAllocationType_kCPU: ..`, and make it a `cdef` instead of `def` (not entirely sure this is worth it)